pilot/adapter/config/file/monitor.go

```diff
@@ -0,0 +1,245 @@
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"reflect"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+
+	"istio.io/istio/pilot/adapter/config/crd"
+	"istio.io/istio/pilot/model"
+)
+
+const (
+	defaultDuration = time.Second / 2
+)
+
+var (
+	supportedExtensions = map[string]bool{
+		".yaml": true,
+		".yml":  true,
+	}
+)
+
+// compareIds compares the IDs (i.e. Namespace, Type, and Name) of the two configs and returns
+// 0 if a == b, -1 if a < b, and 1 if a > b. Used for sorting config arrays.
+func compareIds(a, b *model.Config) int {
+	// Compare Namespace
+	if v := strings.Compare(a.Namespace, b.Namespace); v != 0 {
+		return v
+	}
+	// Compare Type
+	if v := strings.Compare(a.Type, b.Type); v != 0 {
+		return v
+	}
+	// Compare Name
+	return strings.Compare(a.Name, b.Name)
+}
+
+// ByKey is an array of config objects that is capable or sorting by Namespace, Type, and Name.
+type ByKey []*model.Config
+
+func (rs ByKey) Len() int {
+	return len(rs)
+}
+
+func (rs ByKey) Swap(i, j int) {
+	rs[i], rs[j] = rs[j], rs[i]
+}
+
+func (rs ByKey) Less(i, j int) bool {
+	return compareIds(rs[i], rs[j]) < 0
+}
+
+// Monitor will monitor a directory containing yaml files for changes, updating a ConfigStore as changes are
+// detected.
+type Monitor struct {
+	store         model.ConfigStore
+	root          string
+	types         map[string]bool
+	checkDuration time.Duration
+	configs       []*model.Config
+}
+
+// NewMonitor creates a new config store that monitors files under the given root directory for changes to config.
+// If no types are provided (nil or empty), all IstioConfigTypes will be allowed.
+func NewMonitor(delegateStore model.ConfigStore, rootDirectory string, types []string) *Monitor {
+	monitor := &Monitor{
+		store:         delegateStore,
+		root:          rootDirectory,
+		types:         map[string]bool{},
+		checkDuration: defaultDuration,
+	}
+
+	if len(types) == 0 {
+		types = model.IstioConfigTypes.Types()
+	}
+
+	for _, k := range types {
+		if schema, ok := model.IstioConfigTypes.GetByType(k); ok {
+			monitor.types[schema.Type] = true
+		}
+	}
+	return monitor
+}
+
+// Start launches a thread that monitors files under the root directory, updating the underlying config
+// store when changes are detected.
+func (m *Monitor) Start(ctx context.Context) {
+	m.checkAndUpdate()
+	tick := time.NewTicker(m.checkDuration)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				tick.Stop()
+				return
+			case <-tick.C:
+				m.checkAndUpdate()
+			}
+		}
+	}()
+}
+
+// parseInputs is identical to crd.ParseInputs, except that it returns an array of config pointers.
+func parseInputs(data []byte) ([]*model.Config, error) {
+	configs, _, err := crd.ParseInputs(string(data))
+
+	// Convert to an array of pointers.
+	configRefs := make([]*model.Config, len(configs))
+	for i, cfg := range configs {
+		// Make a copy so that we can retain a pointer to it without it being overwritten in the next iteration.
```

Comment:
Is this actually required? I'd expect `configRefs[i] = &cfg` to work.

Reply:
The behavior I was seeing was that the resulting list would contain multiple instances of the same configuration.  This seems to imply that the compiler is reusing the same cfg struct in each iteration, which makes sense. So yeah, I think some form of copy is necessary ... not sure if the approach I'm taking is idiomatic, however.

Reply:
[Yup verified you can't use the loop var.](https://play.golang.org/p/3fR3TRL_Mz)

The way to do this without a copy is to get the address from the array itself, and not take the address of the loop variable:
```go
refs := make([]*model.Config, len(configs))
for i := range configs {
    refs[i] = &configs[i]
}
```

Reply:
done.